jtlibrary/python/jtmodules/src/jtmodules/generate_volume_image.py

```diff
@@ -15,30 +15,51 @@
 import collections
 import logging
 import numpy as np
-from jtlib.segmentation import detect_blobs
+import mahotas as mh
+from jtlib.filter import log_2d, log_3d
 
 logger = logging.getLogger(__name__)
 
-VERSION = '0.1.0'
+VERSION = '0.5.0'
 
-Output = collections.namedtuple('Output', ['volume_image', 'figure'])
+Output = collections.namedtuple(
+    'Output', ['volume_image', 'smoothed_surface_image', 'bead_coordinate_image', 'slide_surface_image', 'figure']
+)
+Beads = collections.namedtuple('Beads', ['coordinates', 'coordinate_image'])
 
 
-def array_to_coordinate_list(array):
+class InvalidSlideError(ValueError):
+    '''Raised when the slide surface is computed from a biased
+    set of points'''
+
+
+def array_to_coordinate_list(array, offset=[0,0]):
     '''Convert a 2D array representation of points in 3D
     to a list of x,y,z coordinates'''
-    points = []
-    for ix in range(array.shape[0]):
-        for iy in range(array.shape[1]):
-            if (array[ix, iy] > 0):
-                points.append((ix, iy, array[ix, iy]))
-    return points
+    nonzero = np.nonzero(array)
+    coordinates = np.vstack([
+        np.stack([nonzero[0] + offset[0], nonzero[1] + offset[1]]),
+        array[nonzero]
+    ]).T
+    return list(map(tuple, coordinates))
+
+
+def coordinate_list_to_array(coordinates, shape, dtype=np.uint16):
+    '''Convert a list of x,y,z coordinates to a 2D array
+    representation of points in 3D'''
+    image = np.zeros(shape, dtype=dtype)
+    for i in range(len(coordinates)):
+        image[coordinates[i][0], coordinates[i][1]] = coordinates[i][2]
```

Comment:
```suggestion
    for coord in coordinates:
        image[coord[0], coord[1]] = coord[2]
```

Reply:
Or, even better:
```suggestion
    for x, y, z in coordinates:
        image[x, y] = z
```
But I don't know if x y z are the "correct" variable names: "esploding" a variable into the components you need is usually better than accessing its values by index.